### DISCUSSION TITLE
Type for $re value should be RegExp | string?

### DISCUSSION THREAD

**Discussion by kamiyo** [mikro-orm/mikro-orm#4589] [General]
**Describe the bug**
In the types for `OperatorMap<T>`, shouldn't the type for `$re` be Regexp | string?

**Expected behavior**
If I'm not mistaken reading through the code, I should be able to pass into a `FilterQuery` a search like:
```
{
    documents: {
        $re: new RegExp('(search|string)', 'i'),
    }
}
```
Especially since I'm wanting to force it to use `~*` (case insensitive) operator for Postgres instead of `~`, and I'm not sure how to do that if I just pass in a normal string. It also doesn't work with RegExp literals `/(search|string)/i`

**Versions**

| Dependency | Version |
| - | - |
| node | 20.4.0 |
| typescript | 5.1.6 |
| mikro-orm | 5.7.13 |
| your-driver | Postgresql |

**Comment by kamiyo** [comment]
Actually something isn't quite behaving correctly. If I force the RegExp with 'i' flag to `any`, it correctly uses `~*`, but the actual SQL generated is something like `where document ~* '/(search|string)/i'`, so it's not really parsing out the regexp. But if I just use a normal string, it uses just the `~` operator.

**Comment by B4nan** [comment]
Nope, `$re` is an explicit operator where you should be passing a string. You can use regexp expression too, but _without_ the operator.

```ts
{
    documents: /(search|string)/i,
}
```

Or you should be able to do this:

```ts
{
    documents: {
        $re: '(search|string)',
        $flags: 'i',
    }
}
```

This issue is related: https://github.com/mikro-orm/mikro-orm/issues/3801

**Reply by kamiyo** [reply]
I didn't realize I could pass a regex string directly, so that's helpful.

If the `$flags` parameter does work,t hen the typescript def needs to be updated then, because it's giving me a ts error.

**Reply by kamiyo** [reply]
Actually, I just tried $flags, and it's not working:
```Error: Trying to query by not existing property Table.$flags```

**Reply by B4nan** [reply]
I think the `$flags` operator might be just an internal thing (a way how the regexp literal gets rewritten), given it's not documented nor tested.
